### ISSUE TITLE
containers slow to launch

### ISSUE THREAD

**Issue by helenkirk** [opencadc/science-platform#268]
I don't think I see this issue already covered here... I have been noticing last week and this week that launching a casa container is  very slow (much slower than last time I was on the system).  I didn't time it to give a precise number, but the launch of casa4.5.0 for me just now was sufficiently long that I got logged out twice (and was approaching the third log out) due to system inactivity.  I don't have anything else running in my session yet, so not an issue of me using too many other resources at the same time.

**Comment by brianmajor** [comment]
Hi Helen!  The length of time to launch a casa container depends on 2 things:  if the container has already been pulled to the registry on the node on which casa was scheduled, the launch should be pretty much immediate.  If not, the node has to pull the image, so it is entirely network dependent then.  If the network in the platform is slow, so will be the time to launch it.  The casa images are also quite large.  There are opportunities to decrease those sizes we could consider doing.

Also, regarding the frequent log-outs in arcade, I think they'll be extended to be over 1 hour after a release on Monday.  So hopefully no more annoyances from that.

**Comment by helenkirk** [comment]
Hi Brian,
Thanks for the speedy reply!  I'm not sure that I entirely understand your first point.  Would it be fair to say that your first point would imply that from the user's perspective, launching a specific version of CASA might be slow the first time but that subsequent launches within the same session would be much faster?  (I can tell there is more nuance to your point than that, but am trying to translate in terms of a typical non-expert astronomer perspective)

Great to hear about the longer time-outs.  In this particular instance, it wasn't all that annoying... I mentioned it just to give some kind of measure of how long the wait was (since you'd probably know what the time-out length was).

**Comment by brianmajor** [comment]
Yes, let me try to explain that.  There are some number of nodes (currently 12) on which you casa container will be scheduled.  There are some tendencies to where it would be scheduled but can be thought of mostly random.  Each of those nodes has a local registry to store previously run containers.  So yes: if you (or someone else) launched a particular casa container that was scheduled on the same node on which your current casa container was scheduled, it would be fast.  Otherwise, the node will need to pull it (then save it locally) before it appears.

So, over time the nodes have larger "caches" of different containers, so will make containers faster to launch in general.  But, now and then the nodes are restarted and lose that cache, so need to start over.

I believe there are ways to 'prime' the nodes with containers so that they are present immediately, but there'd probably be some drawbacks of that approach too.

**Comment by helenkirk** [comment]
Ah, thanks Brian.  I think I understand better now.  From a user perspective then, it's going to be completely random whether or not a given casa container is slow to launch, but the odds are somewhat better that it'll be fast if they've recently launched others of the same version (i.e., they'd have at least the 1/12 chance it would be fast).

Is the system keeping track of which CASA containers are called (i.e., # of calls per CASA version)?  That could be very useful information - e.g., if reducing the container size ends up not working (or not speeding things up enough), perhaps a next step to investigate would be priming the nodes with the two or three most popular casa versions?

**Comment by brianmajor** [comment]
Yes, from a user's perspective, pretty much seems random.

Good question about keeping track of container use.  I just logged in to the image registry and sure enough there are logs of every image download (and upload).  There have been 1 026 452 of them so far!  I expect that list could be exported and sorted to find out what version of CASA is most popular.  Good idea.

As for the image sizes: they're not actually as big as I thought:  between 1.5 and 2.5 Gb.  Still, making them lean would help wit the network transfer.

**Comment by helenkirk** [comment]
Wow!  That's a huge number of image downloads!  I'm glad that info is being gathered somewhere without any effort required at this stage!

Would making the images leaner be something that needs astronomer help, or would it be mostly on the technical side of things?

**Comment by brianmajor** [comment]
Yes, it really is a big number.  We have a 'headless' mode (a pseudo-batch) which is probably a significant portion of those downloads.

It's pretty technical.  For example, with these set of casa versions, it's a matter of refactoring the Dockerfile to be equivalent but more efficient.  Maybe some things can be removed.  Maybe a smaller base image can be used.  I am not an expert in this area, but have heard that combining more steps into a single RUN directive helps, because Docker creates a 'layer' for each directive.

https://github.com/opencadc/skaha/tree/master/containers/software-containers/arcade-casa/version-6.1-6.4

**Comment by sfabbro** [comment]
@brianmajor  - Do you know how much pre-pulling of the container images is done and there is much room for improvement?

**Comment by brianmajor** [comment]
@sfabbro  No pre-pulling is done now.  There's only caching of images that have been pulled to the node already.
